python/ray/dag/tests/experimental/test_torch_tensor_dag.py

```diff
@@ -1172,6 +1172,362 @@ def destroy(self) -> None:
+    shape = (10,)
+    dtype = torch.float16
+    ref = compiled_dag.execute([(shape, dtype, i + 1) for i in range(num_workers)])
+    result = ray.get(ref)
+    reduced_val = ((num_workers + 1) * num_workers / 2) * 2
+    assert result == [(reduced_val, shape, dtype) for _ in workers]
+
+    compiled_dag.teardown()
+
+
+@pytest.mark.parametrize("ray_start_regular", [{"num_cpus": 4}], indirect=True)
+def test_torch_tensor_nccl_comm_deduplication_collective_equals_p2p(ray_start_regular):
+    """
+    Test communicators are deduplicated when the collective and the p2p send/recv
+    are on the same set of actors.
+    """
+    if not USE_GPU:
+        pytest.skip("NCCL tests require GPUs")
+
+    assert (
+        sum(node["Resources"].get("GPU", 0) for node in ray.nodes()) > 1
+    ), "This test requires at least 2 GPUs"
+
+    actor_cls = TorchTensorWorker.options(num_cpus=0, num_gpus=1)
+
+    num_workers = 2
+    workers = [actor_cls.remote() for _ in range(num_workers)]
+
+    with InputNode() as inp:
+        computes = [
+            worker.compute_with_tuple_args.bind(inp, i)
+            for i, worker in enumerate(workers)
+        ]
+        collectives = collective.allreduce.bind(computes)
+        recvs = [
+            worker.recv.bind(
+                collective.with_type_hint(TorchTensorType(transport="nccl"))
+            )
+            for worker, collective in zip(workers, collectives)
+        ]
+        dag = MultiOutputNode(recvs)
+
+    compiled_dag = dag.experimental_compile()
+
+    from ray.dag.collective_node import CollectiveOutputNode
+
+    comms = set()
+    for _, exec_tasks in compiled_dag.actor_to_executable_tasks.items():
+        for exec_task in exec_tasks:
+            dag_node = compiled_dag.idx_to_task[exec_task.task_idx].dag_node
+            if isinstance(dag_node, CollectiveOutputNode):
+                assert exec_task.collective_group is not None
+                nccl_group_id = exec_task.collective_group._type_hint._nccl_group_id
+                assert nccl_group_id is not None
+                comms.add(nccl_group_id)
+    # Only 1 NCCL group should be created.
+    assert len(comms) == 1
+    comm = list(comms)[0]
+    # The comm for all-reduce should be the same as the p2p send/recv comm.
+    assert comm == compiled_dag._nccl_group_id
+
+    # Sanity check: the compiled dag can execute.
+    shape = (10,)
+    dtype = torch.float16
+    ref = compiled_dag.execute([(shape, dtype, i + 1) for i in range(num_workers)])
+    result = ray.get(ref)
+    reduced_val = (num_workers + 1) * num_workers / 2
+    assert result == [(reduced_val, shape, dtype) for _ in workers]
+
+    compiled_dag.teardown()
+
+    receiver = workers[1]
+    with InputNode() as inp:
+        computes = [
+            worker.compute_with_tuple_args.bind(inp, i)
+            for i, worker in enumerate(workers)
+        ]
+        collectives = collective.allreduce.bind(computes)
+        dag = receiver.recv.bind(
```

Comment:
You can directly write `workers[1].recv.bind()` instead of `receiver` since `receiver` is only used once.
Add a comment that "Sender is workers[0] and receiver is workers[1].".

Reply:
Done